app/forms/firewall-rules-common.tsx

```diff
@@ -6,115 +6,328 @@
+      size="sm"
+      className="mr-2.5"
+      disabled={!isDirty}
+      onClick={onClear}
+    >
+      Clear
+    </Button>
+    <Button size="sm" onClick={onSubmit}>
+      {buttonCopy}
+    </Button>
+  </div>
+)
+
+type TypeAndValueTableProps = {
+  sectionType: 'target' | 'host'
+  items: ControllerRenderProps<FirewallRuleValues, 'targets' | 'hosts'>
+}
+const TypeAndValueTable = ({ sectionType, items }: TypeAndValueTableProps) => (
+  <MiniTable.Table
+    className="mb-4"
+    aria-label={sectionType === 'target' ? 'Targets' : 'Host filters'}
+  >
+    <MiniTable.Header>
+      <MiniTable.HeadCell>Type</MiniTable.HeadCell>
+      <MiniTable.HeadCell>Value</MiniTable.HeadCell>
+      {/* For remove button */}
+      <MiniTable.HeadCell className="w-12" />
+    </MiniTable.Header>
+    <MiniTable.Body>
+      {items.value.map(({ type, value }, index) => (
+        <MiniTable.Row
+          tabIndex={0}
+          aria-rowindex={index + 1}
+          aria-label={`Name: ${value}, Type: ${type}`}
+          key={`${type}|${value}`}
         >
-          API docs
-        </a>{' '}
-        to learn more about firewall rules.
-      </>
-    }
-  />
+          <MiniTable.Cell>
+            <Badge variant="solid">{type}</Badge>
+          </MiniTable.Cell>
+          <MiniTable.Cell>{value}</MiniTable.Cell>
+          <MiniTable.RemoveCell
+            onClick={() =>
+              items.onChange(
+                items.value.filter((i) => !(i.value === value && i.type === type))
+              )
+            }
+            label={`remove ${sectionType} ${value}`}
+          />
+        </MiniTable.Row>
+      ))}
+    </MiniTable.Body>
+  </MiniTable.Table>
+)
+
+// Given an array of committed items (VPCs, Subnets, Instances) and
+// a list of all items, return the items that are available
+const availableItems = (
+  committedItems: Array<VpcFirewallRuleTarget | VpcFirewallRuleHostFilter>,
+  // Items is conditional because VPC Subnet fetching isn't 100% guaranteed
+  items?: Array<Vpc | VpcSubnet | Instance>
+) => {
+  if (!items) return []
+  return items
+    .map((item) => item.name)
+    .filter((name) => !committedItems.map((ci) => ci.value).includes(name))
+    .map((name) => toComboboxItem(name))
+}
+
+type ProtocolFieldProps = {
+  control: Control<FirewallRuleValues>
+  protocol: 'TCP' | 'UDP' | 'ICMP'
+}
+const ProtocolField = ({ control, protocol }: ProtocolFieldProps) => (
+  <div>
+    <CheckboxField name="protocols" value={protocol} control={control}>
+      {protocol}
+    </CheckboxField>
+  </div>
 )
 
-export const CommonFields = ({ error, control, nameTaken }: CommonFieldsProps) => {
-  const portRangeForm = useForm({ defaultValues: portRangeDefaultValues })
+type CommonFieldsProps = {
+  control: Control<FirewallRuleValues>
+  project: string
+  instances: Array<Instance>
+  vpcs: Array<Vpc>
+  nameTaken: (name: string) => boolean
+  error: ApiError | null
+}
+
+export const CommonFields = ({
+  control,
+  project,
```

Comment:
I'd just `const { project } = useProjectSelector()` in here and avoid this prop.